vulnfeeds/cpp/main.go

```diff
@@ -206,63 +208,133 @@ func ReposForCPE(CVE string, cache VendorProductToRepoMap, vp VendorProduct, ref
+		return fmt.Errorf("failed to create dir: %v", err)
 	}
 	outputFile := filepath.Join(vulnDir, v.ID+extension)
 	notesFile := filepath.Join(vulnDir, v.ID+".notes")
 	f, err := os.Create(outputFile)
 	if err != nil {
 		Logger.Warnf("Failed to open %s for writing: %v", outputFile, err)
-		return fmt.Errorf("Failed to open %s for writing: %v", outputFile, err)
+		return fmt.Errorf("failed to open %s for writing: %v", outputFile, err)
 	}
 	defer f.Close()
 	err = v.ToJSON(f)
 	if err != nil {
 		Logger.Warnf("Failed to write %s: %v", outputFile, err)
-		return fmt.Errorf("Failed to write %s: %v", outputFile, err)
+		return fmt.Errorf("failed to write %s: %v", outputFile, err)
 	}
-	Logger.Infof("[%s]: Generated OSV record from for %q", CVE.CVE.CVEDataMeta.ID, CPE.Product)
+	Logger.Infof("[%s]: Generated OSV record from for %q", CVEID, CPE.Product)
 	if len(notes) > 0 {
 		err = os.WriteFile(notesFile, []byte(strings.Join(notes, "\n")), 0660)
 		if err != nil {
-			Logger.Warnf("[%s]: Failed to write %s: %v", CVE.CVE.CVEDataMeta.ID, notesFile, err)
+			Logger.Warnf("[%s]: Failed to write %s: %v", CVEID, notesFile, err)
 		}
 	}
 	return nil
 }
 
+// Takes an NVD CVE record and outputs a PackageInfo struct in a file in the specified directory.
+func CVEToPackageInfo(CVE cves.CVEItem, repos []string, cache git.RepoTagsCache, directory string) error {
+	CVEID := CVE.CVE.CVEDataMeta.ID // For brevity.
+	CPEs := cves.CPEs(CVE)
+	CPE, err := cves.ParseCPE(CPEs[0]) // For naming the subdirectory used for output.
+	if err != nil {
+		return fmt.Errorf("[%s]: Can't generate an OSV record without valid CPE data", CVEID)
+	}
+
+	// more often than not, this yields a VersionInfo with AffectedVersions and no AffectedCommits.
+	versions, notes := cves.ExtractVersionInfo(CVE, nil)
+
+	if len(versions.AffectedVersions) != 0 {
+		// There are some AffectedVersions to try and resolve to AffectedCommits.
+		if len(repos) == 0 {
+			return fmt.Errorf("[%s]: No affected ranges for %q, and no repos to try and convert %+v to tags with", CVEID, CPE.Product, versions.AffectedVersions)
+		}
+		Logger.Infof("[%s]: Trying to convert version tags %+v to commits using %v", CVEID, versions.AffectedVersions, repos)
+		versions, err = GitVersionsToCommits(CVEID, versions, repos, cache)
+		if err != nil {
+			return fmt.Errorf("[%s]: Failed to convert version tags to commits: %#v", CVEID, err)
+		}
+	}
+
+	if len(versions.AffectedCommits) == 0 {
+		return fmt.Errorf("[%s]: No affected commit ranges determined for %q", CVEID, CPE.Product)
+	}
+
+	pi := vulns.PackageInfo{VersionInfo: versions}
+
+	vulnDir := filepath.Join(directory, CPE.Vendor, CPE.Product)
+	err = os.MkdirAll(vulnDir, 0755)
+	if err != nil {
+		Logger.Warnf("Failed to create dir: %v", err)
+		return fmt.Errorf("failed to create dir: %v", err)
+	}
+
+	outputFile := filepath.Join(vulnDir, CVEID+".nvd"+extension)
+	notesFile := filepath.Join(vulnDir, CVEID+".nvd.notes")
```

Comment:
Question: Are notes actually needed for this use case?

Reply:
> Question: Are notes actually needed for this use case?

Probably not in the long run, I'm keeping them around for now so they're available for manual human inspection (like when I'm running things locally) but I can see them getting removed in the near future.